store/bigtable/bigtable.go

```diff
@@ -0,0 +1,470 @@
+	// metric store.bigtable.byte_size.at_save is the number of chunk bytes saved in each bulkApply
+	chunkBytesAtSave = stats.NewMeter32("store.bigtable.chunk_bytes.at_save", true)
+)
+
+func formatRowKey(key schema.AMKey, ts uint32) string {
+	return fmt.Sprintf("%s_%d", key.MKey.String(), ts/Month_sec)
+}
+
+func formatFamily(ttl uint32) string {
+	return fmt.Sprintf("ret_%d", ttlToHours(ttl))
+}
+
+func PrepareChunkData(span uint32, data []byte) []byte {
+	chunkSizeAtSave.Value(len(data))
+	version := chunk.FormatStandardGoTszWithSpan
+	buf := new(bytes.Buffer)
+	binary.Write(buf, binary.LittleEndian, version)
+
+	spanCode, ok := chunk.RevChunkSpans[span]
+	if !ok {
+		// it's probably better to panic than to persist the chunk with a wrong length
+		panic(fmt.Sprintf("Chunk span invalid: %d", span))
+	}
+	binary.Write(buf, binary.LittleEndian, spanCode)
+	buf.Write(data)
+	return buf.Bytes()
+}
+
+func mutationFromWriteRequest(cwr *mdata.ChunkWriteRequest) (*bigtable.Mutation, int) {
+	mut := bigtable.NewMutation()
+	family := formatFamily(cwr.TTL)
+	column := "raw"
+	parts := strings.SplitN(cwr.Key.String(), "_", 2)
+	if len(parts) > 1 {
+		column = parts[1]
+	}
+	value := PrepareChunkData(cwr.Span, cwr.Chunk.Series.Bytes())
+	chunkSizeAtSave.Value(len(value))
+	mut.Set(family, column, bigtable.Timestamp(int64(cwr.Chunk.T0)*1e6), value)
+	return mut, len(value)
+}
+
+func ttlToHours(ttl uint32) uint32 {
+	hours := ttl / (60 * 60)
+	if hours < 1 {
+		hours = 1
+	}
+	return hours
+}
+
+type Store struct {
+	tbl              *bigtable.Table
+	client           *bigtable.Client
+	writeQueues      []chan *mdata.ChunkWriteRequest
+	writeQueueMeters []*stats.Range32
+	readQueue        chan struct{}
+	shutdown         chan struct{}
+	wg               sync.WaitGroup
+	tracer           opentracing.Tracer
+	cfg              *StoreConfig
+}
+
+func NewStore(cfg *StoreConfig, ttls []uint32) (*Store, error) {
+	if cfg.WriteMaxFlushSize >= cfg.WriteQueueSize {
+		return nil, fmt.Errorf("btStore: write-queue-size must be larger then write-max-flush-size")
+	}
+	ctx := context.Background()
+	if cfg.CreateCF {
+		adminClient, err := bigtable.NewAdminClient(ctx, cfg.GcpProject, cfg.BigtableInstance)
+		if err != nil {
+			return nil, fmt.Errorf("btStore: failed to create bigtable admin client. %s", err)
+		}
+		tables, err := adminClient.Tables(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("btStore: failed to list tables. %s", err)
+		}
+		found := false
+		for _, t := range tables {
+			if t == cfg.TableName {
+				found = true
+				break
+			}
+		}
+		if !found {
+			log.Infof("btStore: table %s does not exist. Creating it.", cfg.TableName)
+			table := bigtable.TableConf{
+				TableID:  cfg.TableName,
+				Families: make(map[string]bigtable.GCPolicy),
+			}
+			for _, ttl := range ttls {
+				table.Families[formatFamily(ttl)] = bigtable.MaxAgePolicy(time.Duration(ttl/(60*60)) * time.Hour)
```

Comment:
so ttl is a number in seconds
let's say a user wants to retain data for 2.5 hours
this becomes : `9000 / 3600 * time.Hour` = 2 hours.
seems better to use:
`time.Duration(ttl) * time.Second`